### ISSUE TITLE
Add tests

### ISSUE THREAD

**Issue by cappe987** [cappe987/wiretime#3]
Now that support for software timestamping `-S` has been added we can add some tests. This will need to use veth interfaces.
```
veth1  veth2
  |      |
  +------+
```

This setup can be created by doing the following
```bash
ip link add veth1 type veth peer name veth2
ip link set dev veth1 up
ip link set dev veth2 up
```

```
wiretime --tx veth1 --rx veth2 -S
```

Use tcpdump (or tshark?) to verify packets?

**Comment by Saavrm26** [comment]
Should I make new network namespaces for testing, (maybe put testing configurations like ENV variable, creating and destroying link in a test.sh shell script)

**Comment by cappe987** [comment]
> Should I make new network namespaces for testing, (maybe put testing configurations like ENV variable, creating and destroying link in a test.sh shell script)

It's a good idea to use a network namespace. You can do the setup like
```
ip netns add wiretime
ip link add veth1 netns wiretime type veth peer name veth2 netns wiretime
ip netns exec wiretime ip link set dev veth1 up
ip netns exec wiretime ip link set dev veth2 up
```
And then tear it down after the tests are done.

I'm not entirely sure what would be possible configurations. One possibility is to allow running the tests on hardware as well, for which you would need to specify the ports (and not use the `-S` flag). But feel free to come with ideas.

**Comment by Saavrm26** [comment]
Thanks, I'll look into it.

**Comment by Saavrm26** [comment]
For test.sh I was thinking of putting the setup command there as well as destroying the netns
For example
```bash
#!/bin/bash
ip netns add wiretime
ip link add veth1 netns wiretime type veth peer name veth2 netns wiretime
ip netns exec wiretime ip link set dev veth1 up
ip netns exec wiretime ip link set dev veth2 up
ip netns exec wiretime ifconfig  veth1 172.1.1.2
ip netns exec wiretime ifconfig veth2 172.1.1.3
# commands to start tests with CMAKE with some variables like -DCMAKE_INSTALL_PREFIX=$(pwd) 
# commands to tear stop the veths and destroy the netns
```
By setting ```CMAKE_INSTALL_PREFIX``` to current directory we can avoid running the script in with ```sudo```, as normal ```cmake_install``` will have the default value ```/usr/local```

**Comment by cappe987** [comment]
> For test.sh I was thinking of putting the setup command there as well as destroying the netns For example
> 
> ```shell
> #!/bin/bash
> ip netns add wiretime
> ip link add veth1 netns wiretime type veth peer name veth2 netns wiretime
> ip netns exec wiretime ip link set dev veth1 up
> ip netns exec wiretime ip link set dev veth2 up
> ip netns exec wiretime ifconfig  veth1 172.1.1.2
> ip netns exec wiretime ifconfig veth2 172.1.1.3
> # commands to start tests with CMAKE with some variables like -DCMAKE_INSTALL_PREFIX=$(pwd) 
> # commands to tear stop the veths and destroy the netns
> ```
> 
> By setting `CMAKE_INSTALL_PREFIX` to current directory we can avoid running the script in with `sudo`, as normal `cmake_install` will have the default value `/usr/local`

1. IP addresses are not required because Wiretime operates strictly on layer 2 (MAC addresses). Currently it just uses broadcast.
2. Should the test do a install to the local test directory or something? What did you have in mind? If you have already built it with `make` then you already have the binary. My first thought was a `make test` command. But I'm fine with just having a `./test.sh`, and then it might make sense that it builds it all and runs the test.
3. I think it will be a good idea to use Unshare. Because all the `ip` commands and `wiretime` normally requires sudo as well. Running the command `unshare -r -n bash` starts a shell where you can do these operations without sudo. `-n` creates a new namespaces so that part becomes taken care of there as well. Though when running on hardware it's not possible to use Unshare, and will require sudo.

**Comment by Saavrm26** [comment]
> 2\. Should the test do a install to the local test directory

Yeah, as I think it would be better to have the testing and debug executables on the localdirectory(this is just a personal preference of mine, to not install on ```/usr/local/``` until release)

> 2\. But I'm fine with just having a `./test.sh`, and then it might make sense that it builds it all and runs the test.

Yeah this is what I was thinking, for local development a single script to build and start the tests and cleanup after the tests





> 3\. I think it will be a good idea to use Unshare

I'll play around a bit with unshare. Maybe something will comeup that can be helpful
